File: algorithm/Programmers/LEVEL3/베스트앨범.java
package algorithm.Programmers.LEVEL3;
import java.util.*;

class Solution {

    public static int[] solution(String[] genres, int[] plays) {
        List<Integer> answerList = new ArrayList<>();
        Map<String, List<int[]>> dic = new HashMap<>();
        Map<String, Integer> playSum = new HashMap<>();

        for(int i=0;i<plays.length;i++) {
            String genre = genres[i];
            int play = plays[i];
            
            if(!dic.containsKey(genre)) {
                dic.put(genre, new ArrayList<>());
                playSum.put(genre, 0);
            }

            dic.get(genre).add(new int[] {i,play});
            playSum.put(genre, playSum.get(genre)+play);
        }

        List<String> sortedGenre = new ArrayList<>(playSum.keySet());
        Collections.sort(sortedGenre, new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                return playSum.get(o2).compareTo(playSum.get(o1));
            }
        });

        for(String genre:sortedGenre) {
            List<int[]> songs = dic.get(genre);
            Collections.sort(songs, new Comparator<int[]>() {
                @Override
                public int compare(int[] o1, int[] o2) {
                    if(o1[1] == o2[1]) {
                        return Integer.compare(o1[0], o2[0]);
                    }
                    return Integer.compare(o2[1],o1[1]);
                }
            });

            if(songs.size() == 1) {
                answerList.add(songs.get(0)[0]);
            } else {
                answerList.add(songs.get(0)[0]);
                answerList.add(songs.get(1)[0]);
            }
        }
        int[] answer = new int[answerList.size()];
        for(int i=0;i<answerList.size();i++) {
            answer[i]=answerList.get(i);
        }
        return answer;
    }

    public static void main(String[] args) {
        int[] answer = solution(new String[] {"classic", "pop", "classic", "classic", "pop"},new int[] {500, 600, 150, 800, 2500});
        System.out.println(Arrays.toString(answer));
    }
}

File: algorithm/Programmers/LEVEL3/베스트앨범.py
# genres[i]: 고유번호가 i인 노래의 장르
# plays[i]: 고유번호가 i인 노래가 재생된 횟수
def solution(genres, plays):
    answer = []
    dic = {}
    for i in range(len(plays)):
        if genres[i] not in dic.keys():
            dic[genres[i]] = [[i,plays[i]]]
        else:
            dic[genres[i]].append([i,plays[i]])

    dic = dict(sorted(dic.items(), key=lambda x:sum([i[1] for i in x[1]]), reverse=True))

    for k in dic:
        b = sorted(dic[k],key=lambda x:(-x[1],x[0]))
        answer.append(b[0][0])
        if len(b) > 1:
            answer.append(b[1][0])

    return answer
